dataRetriever/resolvers/headerResolver.go

```diff
@@ -108,9 +108,20 @@ func (hdrRes *HeaderResolver) resolveHeaderFromNonce(key []byte) ([]byte, error)
 	}
 
 	//Step 2. search the nonce-key pair first in datapool, second in storage
-	value, _ := hdrRes.hdrNonces.Get(nonce)
-	hash, ok := value.([]byte)
-	if hash == nil || !ok {
+	value, ok := hdrRes.hdrNonces.Get(nonce)
+	var hash []byte
+	if ok {
+		value.Range(func(shardId uint32, existingHash []byte) bool {
+			if shardId == hdrRes.TargetShardID() {
+				hash = existingHash
+				return false
+			}
+
+			return true
+		})
+	}
+
+	if hash == nil {
 		hash, err = hdrRes.hdrNoncesStorage.Get(key)
```

Comment:
header resolver to know whether this is new data, people are asking - 2-3 rounds old, before commit to storage, or it is old data. Than it would not test storage for any querry.